IdentityCore/src/network/error_handler/MSIDAADRequestErrorHandler.m

```diff
@@ -47,38 +48,64 @@ - (void)handleError:(NSError * )error
+
+    if (responseObject)
+    {
+        MSIDAADTokenResponse *tokenResponse = [[MSIDAADTokenResponse alloc] initWithJSONDictionary:responseObject error:nil];
+
+        if (![NSString msidIsStringNilOrBlank:tokenResponse.error])
+        {
+            NSError *oauthError = MSIDCreateError(MSIDOAuthErrorDomain,
+                                                  tokenResponse.oauthErrorCode,
+                                                  tokenResponse.errorDescription,
+                                                  tokenResponse.error,
+                                                  tokenResponse.suberror,
+                                                  nil,
+                                                  context.correlationId,
+                                                  nil);
+
+            NSString *message = [NSString stringWithFormat:@"Oauth error raised: %@, sub error: %@, correlation ID: %@", tokenResponse.error, tokenResponse.suberror, tokenResponse.correlationId];
+            NSString *messagePII = [NSString stringWithFormat:@"Oauth error raised: %@, sub error: %@, correlation ID: %@, description: %@", tokenResponse.error, tokenResponse.suberror, tokenResponse.correlationId, tokenResponse.errorDescription];
+
+            MSID_LOG_WARN(context, @"%@", message);
+            MSID_LOG_WARN_PII(context, @"%@", messagePII);
+
+            if (completionBlock) completionBlock(nil, oauthError);
+            return;
+        }
+    }
+
+    id errorDescription = [NSHTTPURLResponse localizedStringForStatusCode:httpResponse.statusCode];
+
+    NSString *message = [NSString stringWithFormat:@"Http error raised: Http Code: %ld \n", (long)httpResponse.statusCode];
+    NSString *messagePII = [NSString stringWithFormat:@"Http error raised: Http Code: %ld \n%@", (long)httpResponse.statusCode, errorDescription];
     
     MSID_LOG_WARN(context, @"%@", message);
     MSID_LOG_VERBOSE_PII(context, @"%@", messagePII);
     
     NSMutableDictionary *additionalInfo = [NSMutableDictionary new];
-    [additionalInfo setValue:httpResponse.allHeaderFields forKey:MSIDHTTPHeadersKey];
-    [additionalInfo setValue:[NSString stringWithFormat: @"%ld", (long)httpResponse.statusCode]          forKey:MSIDHTTPResponseCodeKey];
+    [additionalInfo setValue:httpResponse.allHeaderFields
+                      forKey:MSIDHTTPHeadersKey];
+
+    [additionalInfo setValue:[NSString stringWithFormat: @"%ld", (long)httpResponse.statusCode]
+                      forKey:MSIDHTTPResponseCodeKey];
     
-    NSError *httpError = MSIDCreateError(MSIDHttpErrorCodeDomain, MSIDErrorServerUnhandledResponse, errorDescription, oauthError, suberror, nil, context.correlationId, additionalInfo);
+    NSError *httpError = MSIDCreateError(MSIDHttpErrorCodeDomain, MSIDErrorServerUnhandledResponse, errorDescription, nil, nil, nil, context.correlationId, additionalInfo);
```

Comment:
Currently we don't have logic to generate errors like `MSIDErrorServerInvalidScope`, do we need to add it here? :)

Reply:
As discussed offline, it's handled lines 65-89 :)